pkg/kubelet/kubelet.go

```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
 	return containerID, nil
 }
 
-func (kl *Kubelet) syncPod(pod *api.BoundPod, containersInPod dockertools.DockerContainers) error {
+type podContainerChangesSpec struct {
+	StartInfraContainer bool
+	InfraContainerId    dockertools.DockerID
+	ContainersToStart   map[int]empty
+	ContainersToKeep    map[dockertools.DockerID]int
```

Comment:
We do, when we find a container which is changed, and thus we restart infra container. In this case we move 'containersToKeep' to 'containersToStart', which, sadly, has to have different types (if we wan't to even pretend to be efficient-ish). containersToKeep are running Docker containers, while containersToStart are specs of containers to start. Yuck...